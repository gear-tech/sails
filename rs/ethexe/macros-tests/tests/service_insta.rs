use proc_macro2::TokenStream;
use quote::quote;
use sails_macros_core::__gservice_internal as gservice;

#[test]
fn works_with_basics() {
    let input = quote! {
        impl SomeService {
            #[export]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> u32 {
                p1
            }

            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_lifetimes_and_generics() {
    let input = quote! {
        impl<'a, 'b, T, U> SomeService<'a, 'b, T, U>
        where T : Clone,
              U: Iterator<Item = u32> {

            #[export]
            pub fn do_this(&mut self) -> u32 {
                42
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_extends() {
    let args = quote! {
        extends = [ExtendedService1, ExtendedService2],
        //arg42 = "Hello, World!"
    };
    let input = quote! {
        impl SomeService {
            #[export]
            pub fn do_this(&mut self) -> u32 {
                42
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_events() {
    let args = quote! {
        events = MyEvents,
    };
    let input = quote! {
        impl MyServiceWithEvents {
            #[export]
            pub fn do_this(&mut self) -> u32 {
                42
            }

            #[export]
            pub fn this(&self) -> bool {
                true
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_lifetimes_and_events() {
    let args = quote! {
        events = MyEvents,
    };
    let input = quote! {
        impl<'a, T> MyGenericEventsService<'a, T>
        where
            T: Clone,
        {
            #[export]
            pub fn do_this(&mut self) -> u32 {
                42
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_extends_and_lifetimes() {
    let args = quote! {
        extends = [BaseWithLifetime<'a>]
    };
    let input = quote! {
        impl<'a> ExtendedWithLifetime<'a> {
            #[export]
            pub fn extended_name(&self) -> String {
                "extended-name".to_string()
            }

            #[export]
            pub fn name(&self) -> String {
                "extended".to_string()
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_methods_with_lifetimes() {
    let input = quote! {
        impl ReferenceService {
            #[export]
            pub fn baked(&self) -> &'static str {
                "Static str!"
            }

            #[export]
            pub fn incr(&mut self) -> &'static ReferenceCount {
                unsafe {
                    COUNTER.0 += 1;
                    &*ptr::addr_of!(COUNTER)
                }
            }

            #[export]
            pub fn add_byte(&mut self, byte: u8) -> &'static [u8] {
                unsafe {
                    BYTES.push(byte);
                    &*ptr::addr_of!(BYTES)
                }
            }

            #[export]
            pub async fn first_byte<'a>(&self) -> Option<&'a u8> {
                unsafe { BYTES.first() }
            }

            #[export]
            pub async fn last_byte<'a>(&self) -> Option<&'a u8> {
                unsafe { BYTES.last() }
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_crate_path() {
    let args = quote!(crate = sails_rename,);
    let input = quote! {
        impl SomeService {
            #[export]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> u32 {
                p1
            }

            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_reply_with_value() {
    let input = quote! {
        impl SomeService {
            #[export]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> CommandReply<u32> {
                (p1, 100_000_000_000).into()
            }

            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_allow_attrs() {
    let input = quote! {
        #[allow(warnings)]
        #[allow(clippy::all)]
        impl SomeService {
            #[allow(unused)]
            #[export]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> u32 {
                p1
            }

            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_docs() {
    let input = quote! {
        impl SomeService {
            /// `DoThis` command
            /// Second line
            #[export]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> u32 {
                p1
            }

            /// `This` query
            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_special_lifetimes_and_events() {
    let args = quote! {
        events = MyEvents,
    };
    let input = quote! {
        impl<T> MyGenericEventsService<'_, '_, T>
        where
            T: Clone,
        {
            #[export]
            pub fn do_this(&mut self) -> u32 {
                42
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_export() {
    let input = quote! {
        impl SomeService {
            #[export(route = "DoSomething", unwrap_result)]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> Result<(u32, String), String> {
                Ok((p1, p2))
            }

            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}
