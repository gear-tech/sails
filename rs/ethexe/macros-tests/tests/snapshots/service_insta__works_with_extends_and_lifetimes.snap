---
source: macros-tests/tests/service_insta.rs
expression: result
---
pub struct ExtendedWithLifetimeExposure<T> {
    route: &'static [u8],
    inner: T,
}
impl<T: sails_rs::meta::ServiceMeta> sails_rs::gstd::services::Exposure
for ExtendedWithLifetimeExposure<T> {
    fn route(&self) -> &'static [u8] {
        self.route
    }
    fn check_asyncness(input: &[u8]) -> Option<bool> {
        use sails_rs::gstd::InvocationIo;
        use sails_rs::gstd::services::{Service, Exposure};
        if !T::ASYNC {
            return Some(false);
        }
        if let Ok(is_async) = extended_with_lifetime_meta::__ExtendedNameParams::check_asyncness(
            input,
        ) {
            return Some(is_async);
        }
        if let Ok(is_async) = extended_with_lifetime_meta::__NameParams::check_asyncness(
            input,
        ) {
            return Some(is_async);
        }
        if let Some(is_async) = <<BaseWithLifetime as Service>::Exposure as Exposure>::check_asyncness(
            input,
        ) {
            return Some(is_async);
        }
        None
    }
}
impl<T> core::ops::Deref for ExtendedWithLifetimeExposure<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}
impl<T> core::ops::DerefMut for ExtendedWithLifetimeExposure<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}
#[allow(clippy::from_over_into)]
impl<'a> Into<<BaseWithLifetime<'a> as sails_rs::gstd::services::Service>::Exposure>
for ExtendedWithLifetimeExposure<ExtendedWithLifetime<'a>> {
    fn into(
        self,
    ) -> <BaseWithLifetime<'a> as sails_rs::gstd::services::Service>::Exposure {
        use sails_rs::gstd::services::Service;
        let base_service: BaseWithLifetime<'a> = self.inner.into();
        base_service.expose(self.route)
    }
}
impl<'a> ExtendedWithLifetimeExposure<ExtendedWithLifetime<'a>> {
    #[export]
    pub fn extended_name(&self) -> String {
        "extended-name".to_string()
    }
    #[export]
    pub fn name(&self) -> String {
        "extended".to_string()
    }
    pub fn check_asyncness(&self, input: &[u8]) -> Option<bool> {
        <Self as sails_rs::gstd::services::Exposure>::check_asyncness(input)
    }
    pub fn try_handle(
        mut self,
        input: &[u8],
        result_handler: fn(&[u8], u128),
    ) -> Option<()> {
        use sails_rs::gstd::InvocationIo;
        use sails_rs::gstd::services::{Service, Exposure};
        if let Ok(request) = extended_with_lifetime_meta::__ExtendedNameParams::decode_params(
            input,
        ) {
            let result = self.extended_name();
            let value = 0u128;
            if sails_rs::prelude::Encode::encoded_size(&result) != 0 {
                extended_with_lifetime_meta::__ExtendedNameParams::with_optimized_encode(
                    &result,
                    self.route().as_ref(),
                    |encoded_result| result_handler(encoded_result, value),
                );
            }
            return Some(());
        }
        if let Ok(request) = extended_with_lifetime_meta::__NameParams::decode_params(
            input,
        ) {
            let result = self.name();
            let value = 0u128;
            if sails_rs::prelude::Encode::encoded_size(&result) != 0 {
                extended_with_lifetime_meta::__NameParams::with_optimized_encode(
                    &result,
                    self.route().as_ref(),
                    |encoded_result| result_handler(encoded_result, value),
                );
            }
            return Some(());
        }
        let base_services: (BaseWithLifetime<'a>) = self.inner.into();
        if base_services.expose(self.route).try_handle(input, result_handler).is_some() {
            return Some(());
        }
        None
    }
    pub async fn try_handle_async(
        mut self,
        input: &[u8],
        result_handler: fn(&[u8], u128),
    ) -> Option<()> {
        use sails_rs::gstd::InvocationIo;
        use sails_rs::gstd::services::{Service, Exposure};
        let base_services: (BaseWithLifetime<'a>) = self.inner.into();
        if base_services
            .expose(self.route)
            .try_handle_async(input, result_handler)
            .await
            .is_some()
        {
            return Some(());
        }
        None
    }
    pub fn try_handle_solidity(
        mut self,
        method: &[u8],
        input: &[u8],
    ) -> Option<(sails_rs::Vec<u8>, u128, bool)> {
        use sails_rs::gstd::services::{Service, Exposure};
        if method
            == &[
                48u8, 69u8, 120u8, 116u8, 101u8, 110u8, 100u8, 101u8, 100u8, 78u8, 97u8,
                109u8, 101u8,
            ]
        {
            let (_, _encode_reply): (u128, bool) = sails_rs::alloy_sol_types::SolValue::abi_decode_params(
                    input,
                    false,
                )
                .ok()?;
            let result = self.extended_name();
            let value = 0u128;
            let output = if _encode_reply {
                let message_id = sails_rs::alloy_primitives::B256::new(
                    sails_rs::gstd::Syscall::message_id().into_bytes(),
                );
                sails_rs::alloy_sol_types::SolValue::abi_encode_sequence(
                    &(message_id, result),
                )
            } else {
                sails_rs::alloy_sol_types::SolValue::abi_encode_sequence(&(result,))
            };
            return Some((output, value, _encode_reply));
        }
        if method == &[16u8, 78u8, 97u8, 109u8, 101u8] {
            let (_, _encode_reply): (u128, bool) = sails_rs::alloy_sol_types::SolValue::abi_decode_params(
                    input,
                    false,
                )
                .ok()?;
            let result = self.name();
            let value = 0u128;
            let output = if _encode_reply {
                let message_id = sails_rs::alloy_primitives::B256::new(
                    sails_rs::gstd::Syscall::message_id().into_bytes(),
                );
                sails_rs::alloy_sol_types::SolValue::abi_encode_sequence(
                    &(message_id, result),
                )
            } else {
                sails_rs::alloy_sol_types::SolValue::abi_encode_sequence(&(result,))
            };
            return Some((output, value, _encode_reply));
        }
        let base_services: (BaseWithLifetime<'a>) = self.inner.into();
        if let Some(result) = base_services
            .expose(self.route)
            .try_handle_solidity(method, input)
        {
            return Some(result);
        }
        None
    }
    pub async fn try_handle_solidity_async(
        mut self,
        method: &[u8],
        input: &[u8],
    ) -> Option<(sails_rs::Vec<u8>, u128, bool)> {
        use sails_rs::gstd::services::{Service, Exposure};
        let base_services: (BaseWithLifetime<'a>) = self.inner.into();
        if let Some(result) = base_services
            .expose(self.route)
            .try_handle_solidity_async(method, input)
            .await
        {
            return Some(result);
        }
        None
    }
}
impl<'a> sails_rs::gstd::services::Service for ExtendedWithLifetime<'a> {
    type Exposure = ExtendedWithLifetimeExposure<Self>;
    fn expose(self, route: &'static [u8]) -> Self::Exposure {
        Self::Exposure {
            route,
            inner: self,
        }
    }
}
impl<'a> sails_rs::meta::ServiceMeta for ExtendedWithLifetime<'a> {
    type CommandsMeta = extended_with_lifetime_meta::CommandsMeta;
    type QueriesMeta = extended_with_lifetime_meta::QueriesMeta;
    type EventsMeta = extended_with_lifetime_meta::EventsMeta;
    const BASE_SERVICES: &'static [sails_rs::meta::AnyServiceMetaFn] = &[
        sails_rs::meta::AnyServiceMeta::new::<BaseWithLifetime>,
    ];
    const ASYNC: bool = <BaseWithLifetime as sails_rs::meta::ServiceMeta>::ASYNC;
}
mod extended_with_lifetime_meta {
    use super::*;
    use sails_rs::{Decode, TypeInfo};
    use sails_rs::gstd::InvocationIo;
    #[derive(Decode, TypeInfo)]
    #[codec(crate = sails_rs::scale_codec)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub struct __ExtendedNameParams {}
    impl InvocationIo for __ExtendedNameParams {
        const ROUTE: &'static [u8] = &[
            48u8, 69u8, 120u8, 116u8, 101u8, 110u8, 100u8, 101u8, 100u8, 78u8, 97u8,
            109u8, 101u8,
        ];
        type Params = Self;
        const ASYNC: bool = false;
    }
    #[derive(Decode, TypeInfo)]
    #[codec(crate = sails_rs::scale_codec)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub struct __NameParams {}
    impl InvocationIo for __NameParams {
        const ROUTE: &'static [u8] = &[16u8, 78u8, 97u8, 109u8, 101u8];
        type Params = Self;
        const ASYNC: bool = false;
    }
    #[derive(TypeInfo)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub enum CommandsMeta {}
    #[derive(TypeInfo)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub enum QueriesMeta {
        ExtendedName(__ExtendedNameParams, String),
        Name(__NameParams, String),
    }
    #[derive(TypeInfo)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub enum NoEvents {}
    pub type EventsMeta = NoEvents;
}
impl<'a> sails_rs::solidity::ServiceSignature for ExtendedWithLifetime<'a> {
    const METHODS: &'static [sails_rs::solidity::MethodExpo] = sails_rs::const_concat_slices!(
        < sails_rs::solidity::MethodExpo >, & [(& [48u8, 69u8, 120u8, 116u8, 101u8,
        110u8, 100u8, 101u8, 100u8, 78u8, 97u8, 109u8, 101u8] as & [u8], "ExtendedName",
        << (u128, bool,) as sails_rs::alloy_sol_types::SolValue > ::SolType as
        sails_rs::alloy_sol_types::SolType > ::SOL_NAME, <<
        (sails_rs::alloy_primitives::B256, String) as sails_rs::alloy_sol_types::SolValue
        > ::SolType as sails_rs::alloy_sol_types::SolType > ::SOL_NAME,), (& [16u8, 78u8,
        97u8, 109u8, 101u8] as & [u8], "Name", << (u128, bool,) as
        sails_rs::alloy_sol_types::SolValue > ::SolType as
        sails_rs::alloy_sol_types::SolType > ::SOL_NAME, <<
        (sails_rs::alloy_primitives::B256, String) as sails_rs::alloy_sol_types::SolValue
        > ::SolType as sails_rs::alloy_sol_types::SolType > ::SOL_NAME,),], <
        BaseWithLifetime as sails_rs::solidity::ServiceSignature > ::METHODS
    );
}
